server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTests.java

```diff
@@ -0,0 +1,362 @@
+
+        assertThat(publication.pendingPublications.keySet(), equalTo(discoNodes));
+        assertTrue(publication.pendingCommits.isEmpty());
+        publication.pendingPublications.entrySet().stream().collect(shuffle()).forEach(e -> {
+            PublishResponse publishResponse = nodeResolver.apply(e.getKey()).coordinationState.handlePublishRequest(
+                publication.publishRequest);
+            e.getValue().onResponse(new PublishWithJoinResponse(publishResponse, Optional.empty()));
+        });
+
+        assertThat(publication.pendingCommits.keySet(), equalTo(discoNodes));
+        assertNotNull(publication.applyCommit);
+        assertEquals(publication.applyCommit.getTerm(), publication.publishRequest.getAcceptedState().term());
+        assertEquals(publication.applyCommit.getVersion(), publication.publishRequest.getAcceptedState().version());
+        publication.pendingCommits.entrySet().stream().collect(shuffle()).forEach(e -> {
+            nodeResolver.apply(e.getKey()).coordinationState.handleCommit(publication.applyCommit);
+            e.getValue().onResponse(TransportResponse.Empty.INSTANCE);
+        });
+
+        assertTrue(publication.completed);
+        assertTrue(publication.committed);
+
+        assertThat(ackListener.await(0L, TimeUnit.SECONDS).size(), equalTo(3));
+    }
+
+    public void testClusterStatePublishingWithFaultyNode() throws InterruptedException {
+        VotingConfiguration singleNodeConfig = new VotingConfiguration(Sets.newHashSet(n1.getId()));
+        initializeCluster(singleNodeConfig);
+
+        AssertingAckListener ackListener = new AssertingAckListener(nodes.size());
+        DiscoveryNodes discoveryNodes = DiscoveryNodes.builder().add(n1).add(n2).add(n3).localNodeId(n1.getId()).build();
+
+        boolean failNodeWhenCommitting = randomBoolean();
+        boolean publicationDidNotMakeItToNode2 = randomBoolean();
+        AtomicInteger remainingActions = new AtomicInteger(
+            failNodeWhenCommitting ? (publicationDidNotMakeItToNode2 ? 2 : 3) : 4);
+        int injectFaultAt = randomInt(remainingActions.get() - 1);
+        logger.info("Injecting fault at: {}, failNodeWhenCommitting: {}, publicationDidNotMakeItToNode2: {}", injectFaultAt,
+            failNodeWhenCommitting, publicationDidNotMakeItToNode2);
+
+        Set<DiscoveryNode> initialFaultyNodes =
+            failNodeWhenCommitting == false && remainingActions.decrementAndGet() == injectFaultAt ?
+                Collections.singleton(n2) : Collections.emptySet();
+        MockPublication publication = node1.publish(CoordinationStateTests.clusterState(1L, 2L,
+            discoveryNodes, singleNodeConfig, singleNodeConfig, 42L), ackListener, initialFaultyNodes);
+
+        publication.pendingPublications.entrySet().stream().collect(shuffle()).forEach(e -> {
+            if (failNodeWhenCommitting == false) {
```

Comment:
We could get rid of this `== false` by inverting this statement.

Reply:
this became obsolete by splitting test into two